src/main/java/org/folio/circulation/domain/validation/ItemLimitValidator.java

```diff
@@ -0,0 +1,119 @@
+package org.folio.circulation.domain.validation;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+import static org.folio.circulation.support.Result.ofAsync;
+import static org.folio.circulation.support.Result.succeeded;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.apache.commons.lang3.StringUtils;
+
+import org.folio.circulation.domain.Item;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanAndRelatedRecords;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.domain.policy.LoanPolicyRepository;
+import org.folio.circulation.support.Result;
+import org.folio.circulation.support.ValidationErrorFailure;
+
+public class ItemLimitValidator {
+  private final Function<String, ValidationErrorFailure> itemLimitErrorFunction;
+  private final LoanRepository loanRepository;
+  private final LoanPolicyRepository loanPolicyRepository;
+
+  public ItemLimitValidator(Function<String, ValidationErrorFailure> itemLimitErrorFunction,
+    LoanRepository loanRepository, LoanPolicyRepository loanPolicyRepository) {
+
+    this.itemLimitErrorFunction = itemLimitErrorFunction;
+    this.loanRepository = loanRepository;
+    this.loanPolicyRepository = loanPolicyRepository;
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> refuseWhenItemLimitIsReached(
+    LoanAndRelatedRecords records) {
+
+    Loan loan = records.getLoan();
+    Integer itemLimit = loan.getLoanPolicy().getItemLimit();
+
+    if (itemLimit == null) {
+      return completedFuture(succeeded(records));
+    }
+
+    return loanPolicyRepository.lookupConditions(loan.getItem(), loan.getUser())
```

Comment:
Does this apply the circulation rules a second time to re-determine the loan policy?

Reply:
yes, it does

Reply:
FOLIO has performance expectations for a check out and applying the circulation rules is a relatively expensive operation. 

I don't think the system should do so unless it has to. Why is there a need to re-evaluate the loan policy twice within the same check out?

Reply:
Could you please recommend how can I retrieve circulation rule condition without applying the circulation rule? 
I retrieve it only when item limit is defined. For that case we have to retrieve it to check rule conditions.

Reply:
My understanding is that this change has made it so that when the rules are applied the criteria (conditions?) used to determine the policy are also provided. Is that the case?

If so, I suggest that this information is retained after the first application of the rules is done, in case it is needed for applying a limit. Rather than forgetting it and having to figure it out a second time (which could be different).

This is a trade off between retaining a bit more information throughout the process versus having to apply the rules again (which could be expensive). I personally think the complexity from retaining this information is a better trade off than the potential performance impact.

Reply:
But... if we didn't have policy ID, how would we check that fetched loans match the policy with specific item limit? Could you please clarify?

Reply:
I have just updated logic, now the rules are applied only one one time. Then this information is retained in LoanPolicy and we don't have to apply ruluse a second tyme.

Reply:
Thanks @roman-barannyk